[Android] Adding Annotations may crash the whole app

# Details

There seems to be an error in the `addAnnotation` function on Android which crashed the whole app. This is especially critical as the error cannot be handled from javascript at all. Therefore users will just see the app close without any message.

```
--------- beginning of crash
07-15 10:54:55.588  4647  4647 E AndroidRuntime: FATAL EXCEPTION: main
07-15 10:54:55.588  4647  4647 E AndroidRuntime: Process: com.myappname, PID: 4647
07-15 10:54:55.588  4647  4647 E AndroidRuntime: io.reactivex.exceptions.OnErrorNotImplementedException: Point is not an array of two floats
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at io.reactivex.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:67)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:200)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:252)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:119)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at android.os.Handler.handleCallback(Handler.java:873)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:99)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:193)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at android.app.ActivityThread.main(ActivityThread.java:6669)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at java.lang.reflect.Method.invoke(Native Method)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
07-15 10:54:55.588  4647  4647 E AndroidRuntime: Caused by: java.lang.RuntimeException: Point is not an array of two floats
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at com.pspdfkit.framework.jni.NativeAnnotationManager$CppProxy.native_createAnnotationFromInstantJson(Native Method)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at com.pspdfkit.framework.jni.NativeAnnotationManager$CppProxy.createAnnotationFromInstantJson(SourceFile:190)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at com.pspdfkit.framework.h.createAnnotationFromInstantJson(SourceFile:576)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at com.pspdfkit.views.PdfView.lambda$addAnnotation$0(PdfView.java:500)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at com.pspdfkit.views.-$$Lambda$PdfView$uhxmUT8S3cNY6m3quqBbR2itr8A.accept(Unknown Source:4)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
07-15 10:54:55.588  4647  4647 E AndroidRuntime:        ... 10 mor
```

# Acceptance Criteria

- [ ] Every error is handled gracefully and reported to the higher level js function. Usually errors should result in a rejected Promise.

--- a/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
+++ b/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
@@ -192,8 +192,9 @@ public class ReactPdfViewManager extends ViewGroupManager<PdfView> {
                 }
                 break;
             case COMMAND_REMOVE_ANNOTATION:
-                if (args != null) {
-                    annotationDisposables.add(root.removeAnnotation(args.getMap(0)));
+                if (args != null && args.size() == 2) {
+                    final int requestId = args.getInt(0);
+                    annotationDisposables.add(root.removeAnnotation(requestId, args.getMap(1)));
                 }
                 break;
             case COMMAND_GET_ALL_UNSAVED_ANNOTATIONS:

--- a/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
+++ b/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
@@ -186,8 +186,9 @@ public class ReactPdfViewManager extends ViewGroupManager<PdfView> {
                 }
                 break;
             case COMMAND_ADD_ANNOTATION:
-                if (args != null) {
-                    annotationDisposables.add(root.addAnnotation(args.getMap(0)));
+                if (args != null && args.size() == 2) {
+                    final int requestId = args.getInt(0);
+                    annotationDisposables.add(root.addAnnotation(requestId, args.getMap(1)));
                 }
                 break;
             case COMMAND_REMOVE_ANNOTATION:

--- a/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
+++ b/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
@@ -211,8 +211,9 @@ public class ReactPdfViewManager extends ViewGroupManager<PdfView> {
                 }
                 break;
             case COMMAND_ADD_ANNOTATIONS:
-                if (args != null && args.size() == 1) {
-                    annotationDisposables.add(root.addAnnotations(args.getMap(0)));
+                if (args != null && args.size() == 2) {
+                    final int requestId = args.getInt(0);
+                    annotationDisposables.add(root.addAnnotations(requestId, args.getMap(1)));
                 }
                 break;
             case COMMAND_GET_FORM_FIELD_VALUE:

--- a/android/src/main/java/com/pspdfkit/react/events/PdfViewDataReturnedEvent.java
+++ b/android/src/main/java/com/pspdfkit/react/events/PdfViewDataReturnedEvent.java
@@ -47,6 +47,16 @@ public class PdfViewDataReturnedEvent extends Event<PdfViewDataReturnedEvent> {
         payload = Arguments.makeNativeMap(map);
     }
 
+    public PdfViewDataReturnedEvent(@IdRes int viewId, int requestId, boolean result) {
+        super(viewId);
+        this.requestId = requestId;
+        Map<String, Object> map = new HashMap<>();
+        map.put("requestId", requestId);
+        map.put("result", result);
+
+        payload = Arguments.makeNativeMap(map);
+    }
+
     public PdfViewDataReturnedEvent(@IdRes int viewId, int requestId, @NonNull JSONObject jsonObject) {
         super(viewId);
         this.requestId = requestId;

--- a/android/src/main/java/com/pspdfkit/views/PdfView.java
+++ b/android/src/main/java/com/pspdfkit/views/PdfView.java
@@ -54,6 +54,7 @@ import java.io.File;
 import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.List;
+import java.util.NoSuchElementException;
 import java.util.concurrent.Callable;
 
 import io.reactivex.Completable;
@@ -502,14 +503,11 @@ public class PdfView extends FrameLayout {
             })
             .map(Annotation::toInstantJson)
             .observeOn(AndroidSchedulers.mainThread())
-            .subscribe((instantJson) -> {
-                JSONObject result = new JSONObject();
-                result.put("annotation", new JSONObject(instantJson));
-                eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, result));
-            }, (throwable) -> eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, throwable)));
+            .subscribe((instantJson) -> eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, true)),
+                (throwable) -> eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, throwable)));
     }
 
-    public Disposable removeAnnotation(ReadableMap annotation) {
+    public Disposable removeAnnotation(final int requestId, ReadableMap annotation) {
         return fragmentGetter.take(1).map(PdfFragment::getDocument).subscribeOn(Schedulers.io())
                 .observeOn(AndroidSchedulers.mainThread())
                 .flatMap(pdfDocument -> {
@@ -522,10 +520,23 @@ public class PdfView extends FrameLayout {
                     if (pageIndex == -1 || type == null || name == null) {
                         return Observable.empty();
                     }
+
                     return pdfDocument.getAnnotationProvider().getAllAnnotationsOfType(getTypeFromString(type), pageIndex, 1)
                             .filter(annotationToFilter -> name.equals(annotationToFilter.getName()))
                             .map(filteredAnnotation -> new Pair<>(filteredAnnotation, pdfDocument));
-                }).subscribe(pair -> pair.second.getAnnotationProvider().removeAnnotationFromPage(pair.first));
+                })
+            .firstOrError()
+            .flatMapCompletable(pair -> Completable.fromAction(() -> {
+                pair.second.getAnnotationProvider().removeAnnotationFromPage(pair.first);
+            }))
+            .subscribe(() -> eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, true)), (throwable -> {
+                if (throwable instanceof NoSuchElementException) {
+                    // We didn't find an annotation so return false.
+                    eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, false));
+                } else {
+                    eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, throwable));
+                }
+            }));
     }
 
     public Single<JSONObject> getAllUnsavedAnnotations() {
@@ -550,13 +561,8 @@ public class PdfView extends FrameLayout {
                 DocumentJsonFormatter.importDocumentJson(currentDocument, dataProvider);
             }))
             .observeOn(AndroidSchedulers.mainThread())
-            .subscribe(() -> {
-                JSONObject result = new JSONObject();
-                result.put("annotations", "created");
-                eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, result));
-            }, (throwable) -> {
-                eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, throwable));
-            });
+            .subscribe(() -> eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, true)),
+                (throwable) -> eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, throwable)));
     }
 
     public Disposable getFormFieldValue(final int requestId, @NonNull String formElementName) {

--- a/android/src/main/java/com/pspdfkit/views/PdfView.java
+++ b/android/src/main/java/com/pspdfkit/views/PdfView.java
@@ -56,6 +56,7 @@ import java.util.EnumSet;
 import java.util.List;
 import java.util.concurrent.Callable;
 
+import io.reactivex.Completable;
 import io.reactivex.Observable;
 import io.reactivex.ObservableSource;
 import io.reactivex.Single;
@@ -536,14 +537,21 @@ public class PdfView extends FrameLayout {
                 });
     }
 
-    public Disposable addAnnotations(ReadableMap annotation) {
+    public Disposable addAnnotations(final int requestId, ReadableMap annotation) {
         return fragmentGetter.take(1).map(PdfFragment::getDocument).subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(pdfDocument -> {
-                    JSONObject json = new JSONObject(annotation.toHashMap());
-                    final DataProvider dataProvider = new DocumentJsonDataProvider(json);
-                    DocumentJsonFormatter.importDocumentJson(pdfDocument, dataProvider);
-                });
+            .flatMapCompletable(currentDocument -> Completable.fromAction(() -> {
+                JSONObject json = new JSONObject(annotation.toHashMap());
+                final DataProvider dataProvider = new DocumentJsonDataProvider(json);
+                DocumentJsonFormatter.importDocumentJson(currentDocument, dataProvider);
+            }))
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(() -> {
+                JSONObject result = new JSONObject();
+                result.put("annotations", "created");
+                eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, result));
+            }, (throwable) -> {
+                eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, throwable));
+            });
     }
 
     public Disposable getFormFieldValue(final int requestId, @NonNull String formElementName) {

--- a/android/src/main/java/com/pspdfkit/views/PdfView.java
+++ b/android/src/main/java/com/pspdfkit/views/PdfView.java
@@ -494,14 +494,19 @@ public class PdfView extends FrameLayout {
         return EnumSet.noneOf(AnnotationType.class);
     }
 
-    public Disposable addAnnotation(ReadableMap annotation) {
+    public Disposable addAnnotation(final int requestId, ReadableMap annotation) {
         return fragmentGetter.take(1).map(PdfFragment::getDocument).subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(pdfDocument -> {
-                    JSONObject json = new JSONObject(annotation.toHashMap());
-                    pdfDocument.getAnnotationProvider().createAnnotationFromInstantJson(json.toString());
-                });
-
+            .map(pdfDocument -> {
+                JSONObject json = new JSONObject(annotation.toHashMap());
+                return pdfDocument.getAnnotationProvider().createAnnotationFromInstantJson(json.toString());
+            })
+            .map(Annotation::toInstantJson)
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe((instantJson) -> {
+                JSONObject result = new JSONObject();
+                result.put("annotation", new JSONObject(instantJson));
+                eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, result));
+            }, (throwable) -> eventDispatcher.dispatchEvent(new PdfViewDataReturnedEvent(getId(), requestId, throwable)));
     }
 
     public Disposable removeAnnotation(ReadableMap annotation) {

--- a/android/src/main/java/com/pspdfkit/views/PdfView.java
+++ b/android/src/main/java/com/pspdfkit/views/PdfView.java
@@ -171,47 +171,46 @@ public class PdfView extends FrameLayout {
     }
 
     public void setConfiguration(PdfActivityConfiguration configuration) {
+        if (configuration != null && !configuration.equals(this.configuration)) {
+            // The configuration changed, recreate the fragment.
+            // We set the current page index so the fragment is created at this location.
+            this.pageIndex = fragment != null ? fragment.getPageIndex() : this.pageIndex;
+            removeFragment(false);
+        }
         this.configuration = configuration;
         setupFragment();
     }
 
-    public void setDocument(@Nullable String document) {
-        if (document == null) {
+    public void setDocument(@Nullable String documentPath) {
+        if (documentPath == null) {
+            this.document = null;
             removeFragment(false);
             return;
         }
 
-        if (Uri.parse(document).getScheme() == null) {
+        if (Uri.parse(documentPath).getScheme() == null) {
             // If there is no scheme it might be a raw path.
             try {
-                File file = new File(document);
-                document = Uri.fromFile(file).toString();
+                File file = new File(documentPath);
+                documentPath = Uri.fromFile(file).toString();
             } catch (Exception e) {
-                document = FILE_SCHEME + document;
+                documentPath = FILE_SCHEME + document;
             }
         }
         if (documentOpeningDisposable != null) {
             documentOpeningDisposable.dispose();
         }
         updateState();
-        documentOpeningDisposable = PdfDocumentLoader.openDocumentAsync(getContext(), Uri.parse(document))
+        documentOpeningDisposable = PdfDocumentLoader.openDocumentAsync(getContext(), Uri.parse(documentPath))
                 .subscribeOn(Schedulers.io())
                 .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(new Consumer<PdfDocument>() {
-
-                    @Override
-                    public void accept(PdfDocument pdfDocument) throws Exception {
-                        PdfView.this.document = pdfDocument;
-                        setupFragment();
-                    }
-
-                }, new Consumer<Throwable>() {
-                    @Override
-                    public void accept(Throwable throwable) throws Exception {
-                        PdfView.this.document = null;
-                        setupFragment();
-                        eventDispatcher.dispatchEvent(new PdfViewDocumentLoadFailedEvent(getId(), throwable.getMessage()));
-                    }
+                .subscribe(pdfDocument -> {
+                    PdfView.this.document = pdfDocument;
+                    setupFragment();
+                }, throwable -> {
+                    PdfView.this.document = null;
+                    setupFragment();
+                    eventDispatcher.dispatchEvent(new PdfViewDocumentLoadFailedEvent(getId(), throwable.getMessage()));
                 });
     }
 
@@ -329,15 +328,17 @@ public class PdfView extends FrameLayout {
         PdfFragment pdfFragment = (PdfFragment) fragmentManager.findFragmentByTag(fragmentTag);
         if (pdfFragment != null) {
             fragmentManager.beginTransaction()
-                    .remove(pdfFragment)
-                    .commitAllowingStateLoss();
+                .remove(pdfFragment)
+                .commitNowAllowingStateLoss();
         }
         if (makeInactive) {
+            // Clear everything.
             isActive = false;
+            document = null;
         }
 
         fragment = null;
-        document = null;
+
         fragmentGetter.onComplete();
         fragmentGetter = BehaviorSubject.create();
         pendingFragmentActions.dispose();

--- a/index.js
+++ b/index.js
@@ -256,14 +256,27 @@ class PSPDFKitView extends React.Component {
    * Applies the passed in document instant json.
    *
    * @param annotations The document instant json to apply.
+   * 
+   * Returns a promise resolving a dictionary with the following structure:
+   * {'annotations' : 'created'}.
    */
   addAnnotations = function(annotations) {
     if (Platform.OS === "android") {
+      let requestId = this._nextRequestId++;
+      let requestMap = this._requestMap;
+
+      // We create a promise here that will be resolved once onDataReturned is called.
+      let promise = new Promise(function(resolve, reject) {
+        requestMap[requestId] = { resolve: resolve, reject: reject };
+      });
+
       UIManager.dispatchViewManagerCommand(
         findNodeHandle(this.refs.pdfView),
         this._getViewManagerConfig("RCTPSPDFKitView").Commands.addAnnotations,
-        [annotations]
+        [requestId, annotations]
       );
+
+      return promise
     } else if (Platform.OS === "ios") {
       return NativeModules.PSPDFKitViewManager.addAnnotations(
         annotations,

--- a/index.js
+++ b/index.js
@@ -153,8 +153,7 @@ class PSPDFKitView extends React.Component {
    * @param pageIndex The page to get the annotations for.
    * @param type The type of annotations to get (See here for types https://pspdfkit.com/guides/server/current/api/json-format/) or null to get all annotations.
    *
-   * Returns a promise resolving an array with the following structure:
-   * {'annotations' : [instantJson]}
+   * Returns a promise resolving to true if the annotation was added.
    */
   getAnnotations = function(pageIndex, type) {
     if (Platform.OS === "android") {
@@ -186,9 +185,8 @@ class PSPDFKitView extends React.Component {
    * Adds a new annotation to the current document.
    *
    * @param annotation InstantJson of the annotation to add.
-   * 
-   * Returns a promise resolving a dictionary with the following structure:
-   * {'annotation' : instantJson}.
+   *
+   * Returns a promise resolving to true if the annotation was added.
    */
   addAnnotation = function(annotation) {
     if (Platform.OS === "android") {
@@ -206,7 +204,7 @@ class PSPDFKitView extends React.Component {
         [requestId, annotation]
       );
 
-      return promise
+      return promise;
     } else if (Platform.OS === "ios") {
       return NativeModules.PSPDFKitViewManager.addAnnotation(
         annotation,
@@ -219,14 +217,26 @@ class PSPDFKitView extends React.Component {
    * Removes an existing annotation from the current document.
    *
    * @param annotation InstantJson of the annotation to remove.
+   * 
+   * Returns a promise resolving to true if the annotation was removed and to false if the annotation couldn't be found.
    */
   removeAnnotation = function(annotation) {
     if (Platform.OS === "android") {
+      let requestId = this._nextRequestId++;
+      let requestMap = this._requestMap;
+
+      // We create a promise here that will be resolved once onDataReturned is called.
+      let promise = new Promise(function(resolve, reject) {
+        requestMap[requestId] = { resolve: resolve, reject: reject };
+      });
+
       UIManager.dispatchViewManagerCommand(
         findNodeHandle(this.refs.pdfView),
         this._getViewManagerConfig("RCTPSPDFKitView").Commands.removeAnnotation,
-        [annotation]
+        [requestId, annotation]
       );
+
+      return promise;
     } else if (Platform.OS === "ios") {
       return NativeModules.PSPDFKitViewManager.removeAnnotation(
         annotation,
@@ -269,7 +279,7 @@ class PSPDFKitView extends React.Component {
    * Applies the passed in document instant json.
    *
    * @param annotations The document instant json to apply.
-   * 
+   *
    * Returns a promise resolving a dictionary with the following structure:
    * {'annotations' : 'created'}.
    */
@@ -289,7 +299,7 @@ class PSPDFKitView extends React.Component {
         [requestId, annotations]
       );
 
-      return promise
+      return promise;
     } else if (Platform.OS === "ios") {
       return NativeModules.PSPDFKitViewManager.addAnnotations(
         annotations,

--- a/index.js
+++ b/index.js
@@ -186,14 +186,27 @@ class PSPDFKitView extends React.Component {
    * Adds a new annotation to the current document.
    *
    * @param annotation InstantJson of the annotation to add.
+   * 
+   * Returns a promise resolving a dictionary with the following structure:
+   * {'annotation' : instantJson}.
    */
   addAnnotation = function(annotation) {
     if (Platform.OS === "android") {
+      let requestId = this._nextRequestId++;
+      let requestMap = this._requestMap;
+
+      // We create a promise here that will be resolved once onDataReturned is called.
+      let promise = new Promise(function(resolve, reject) {
+        requestMap[requestId] = { resolve: resolve, reject: reject };
+      });
+
       UIManager.dispatchViewManagerCommand(
         findNodeHandle(this.refs.pdfView),
         this._getViewManagerConfig("RCTPSPDFKitView").Commands.addAnnotation,
-        [annotation]
+        [requestId, annotation]
       );
+
+      return promise
     } else if (Platform.OS === "ios") {
       return NativeModules.PSPDFKitViewManager.addAnnotation(
         annotation,

--- a/index.js
+++ b/index.js
@@ -153,7 +153,8 @@ class PSPDFKitView extends React.Component {
    * @param pageIndex The page to get the annotations for.
    * @param type The type of annotations to get (See here for types https://pspdfkit.com/guides/server/current/api/json-format/) or null to get all annotations.
    *
-   * Returns a promise resolving to true if the annotation was added.
+   * Returns a promise resolving an array with the following structure:
+   * {'annotations' : [instantJson]}
    */
   getAnnotations = function(pageIndex, type) {
     if (Platform.OS === "android") {
@@ -217,7 +218,7 @@ class PSPDFKitView extends React.Component {
    * Removes an existing annotation from the current document.
    *
    * @param annotation InstantJson of the annotation to remove.
-   * 
+   *
    * Returns a promise resolving to true if the annotation was removed and to false if the annotation couldn't be found.
    */
   removeAnnotation = function(annotation) {
@@ -280,8 +281,7 @@ class PSPDFKitView extends React.Component {
    *
    * @param annotations The document instant json to apply.
    *
-   * Returns a promise resolving a dictionary with the following structure:
-   * {'annotations' : 'created'}.
+   * Returns a promise resolving to true if the annotation was added.
    */
   addAnnotations = function(annotations) {
     if (Platform.OS === "android") {

--- a/index.js
+++ b/index.js
@@ -187,7 +187,7 @@ class PSPDFKitView extends React.Component {
    *
    * @param annotation InstantJson of the annotation to add.
    *
-   * Returns a promise resolving to true if the annotation was added.
+   * Returns a promise resolving to true if the annotation was added. Otherwise, returns false if an error has occurred.
    */
   addAnnotation = function(annotation) {
     if (Platform.OS === "android") {
@@ -219,7 +219,7 @@ class PSPDFKitView extends React.Component {
    *
    * @param annotation InstantJson of the annotation to remove.
    *
-   * Returns a promise resolving to true if the annotation was removed and to false if the annotation couldn't be found.
+   * Returns a promise resolving to true if the annotation was removed. Otherwise, returns false if the annotation couldn't be found.
    */
   removeAnnotation = function(annotation) {
     if (Platform.OS === "android") {

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "react-native-pspdfkit",
-  "version": "1.24.8",
+  "version": "1.24.9",
   "description": "A React Native module for the PSPDFKit library.",
   "keywords": [
     "react native",

--- a/samples/Catalog/Catalog.android.js
+++ b/samples/Catalog/Catalog.android.js
@@ -632,7 +632,12 @@ class PdfViewInstantJsonScreen extends Component<{}> {
                   type: "pspdfkit/ink",
                   updatedAt: "2018-07-03T13:53:03Z",
                   v: 1
-                });
+                }).then(result => {
+                  alert(JSON.stringify(result))
+                })
+                .catch(ex => {
+                  alert(JSON.stringify(ex))
+                })
               }}
               title="Add annotation"
             />

--- a/samples/Catalog/Catalog.android.js
+++ b/samples/Catalog/Catalog.android.js
@@ -22,7 +22,7 @@ import {
 } from "react-native";
 import { StackNavigator, NavigationEvents } from "react-navigation";
 
-import QRCodeScanner from 'react-native-qrcode-scanner';
+import QRCodeScanner from "react-native-qrcode-scanner";
 
 import PSPDFKitView from "react-native-pspdfkit";
 
@@ -69,8 +69,8 @@ const examples = [
     name: "Open local document",
     description: "Open document from external storage directory.",
     action: () => {
-      requestExternalStoragePermission(function () {
-        extractFromAssetsIfMissing("Annual Report.pdf", function () {
+      requestExternalStoragePermission(function() {
+        extractFromAssetsIfMissing("Annual Report.pdf", function() {
           PSPDFKit.present(DOCUMENT, {});
         });
       });
@@ -81,8 +81,8 @@ const examples = [
     name: "Open local image document",
     description: "Open image image document from external storage directory.",
     action: () => {
-      requestExternalStoragePermission(function () {
-        extractFromAssetsIfMissing("android.png", function () {
+      requestExternalStoragePermission(function() {
+        extractFromAssetsIfMissing("android.png", function() {
           PSPDFKit.presentImage(IMAGE_DOCUMENT, CONFIGURATION_IMAGE_DOCUMENT);
         });
       });
@@ -94,7 +94,7 @@ const examples = [
     description:
       "You can configure the builder with dictionary representation of the PSPDFConfiguration object.",
     action: () => {
-      requestExternalStoragePermission(function () {
+      requestExternalStoragePermission(function() {
         PSPDFKit.present(DOCUMENT, CONFIGURATION);
       });
     }
@@ -171,7 +171,7 @@ function extractFromAssetsIfMissing(assetFile, callback) {
       } else {
         console.log(
           assetFile +
-          " does not exist, extracting it from assets folder to the external storage directory."
+            " does not exist, extracting it from assets folder to the external storage directory."
         );
         RNFS.existsAssets(assetFile)
           .then(exist => {
@@ -190,7 +190,7 @@ function extractFromAssetsIfMissing(assetFile, callback) {
               // File does not exist, it should never happen.
               throw new Error(
                 assetFile +
-                " couldn't be extracted as it was not found in the project assets folder."
+                  " couldn't be extracted as it was not found in the project assets folder."
               );
             }
           })
@@ -337,7 +337,12 @@ class PdfViewScreen extends Component<{}> {
           }}
           pageIndex={this.state.currentPageIndex}
           fragmentTag="PDF1"
-          menuItemGrouping={['freetext', {key: 'markup', items: ['highlight', "underline"]}, 'ink', 'image']}
+          menuItemGrouping={[
+            "freetext",
+            { key: "markup", items: ["highlight", "underline"] },
+            "ink",
+            "image"
+          ]}
           onStateChanged={event => {
             this.setState({
               currentPageIndex: event.currentPageIndex,
@@ -600,44 +605,51 @@ class PdfViewInstantJsonScreen extends Component<{}> {
             <Button
               onPress={() => {
                 // This adds a new ink annotation to the first page.
-                this.refs.pdfView.addAnnotation({
-                  bbox: [
-                    89.58633422851562,
-                    98.5791015625,
-                    143.12948608398438,
-                    207.1583251953125
-                  ],
-                  blendMode: "normal",
-                  createdAt: "2018-07-03T13:53:03Z",
-                  isDrawnNaturally: false,
-                  lineWidth: 5,
-                  lines: {
-                    intensities: [[0.5, 0.5, 0.5], [0.5, 0.5, 0.5]],
-                    points: [
-                      [
-                        [92.08633422851562, 101.07916259765625],
-                        [92.08633422851562, 202.15826416015625],
-                        [138.12950134277344, 303.2374267578125]
-                      ],
-                      [
-                        [184.17266845703125, 101.07916259765625],
-                        [184.17266845703125, 202.15826416015625],
-                        [230.2158203125, 303.2374267578125]
+                this.refs.pdfView
+                  .addAnnotation({
+                    bbox: [
+                      89.58633422851562,
+                      98.5791015625,
+                      143.12948608398438,
+                      207.1583251953125
+                    ],
+                    blendMode: "normal",
+                    createdAt: "2018-07-03T13:53:03Z",
+                    isDrawnNaturally: false,
+                    lineWidth: 5,
+                    name: 'my annotation',
+                    lines: {
+                      intensities: [[0.5, 0.5, 0.5], [0.5, 0.5, 0.5]],
+                      points: [
+                        [
+                          [92.08633422851562, 101.07916259765625],
+                          [92.08633422851562, 202.15826416015625],
+                          [138.12950134277344, 303.2374267578125]
+                        ],
+                        [
+                          [184.17266845703125, 101.07916259765625],
+                          [184.17266845703125, 202.15826416015625],
+                          [230.2158203125, 303.2374267578125]
+                        ]
                       ]
-                    ]
-                  },
-                  opacity: 1,
-                  pageIndex: 0,
-                  strokeColor: "#AA47BE",
-                  type: "pspdfkit/ink",
-                  updatedAt: "2018-07-03T13:53:03Z",
-                  v: 1
-                }).then(result => {
-                  alert(JSON.stringify(result))
-                })
-                .catch(ex => {
-                  alert(JSON.stringify(ex))
-                })
+                    },
+                    opacity: 1,
+                    pageIndex: 0,
+                    strokeColor: "#AA47BE",
+                    type: "pspdfkit/ink",
+                    updatedAt: "2018-07-03T13:53:03Z",
+                    v: 1
+                  })
+                  .then(result => {
+                    if (result) {
+                      alert("Annotation was added.");
+                    } else {
+                      alert("Annotation was not added.");
+                    }
+                  })
+                  .catch(error => {
+                    alert(JSON.stringify(error));
+                  });
               }}
               title="Add annotation"
             />
@@ -647,10 +659,21 @@ class PdfViewInstantJsonScreen extends Component<{}> {
               onPress={() => {
                 // This removes the first annotation on the first page.
                 this.refs.pdfView.getAnnotations(0, null).then(results => {
-                  const annotations = results.annotations
+                  const annotations = results.annotations;
                   if (annotations.length >= 1) {
                     const annotation = annotations[0];
-                    this.refs.pdfView.removeAnnotation(annotation);
+                    this.refs.pdfView
+                      .removeAnnotation(annotation)
+                      .then(result => {
+                        if (result) {
+                          alert("Annotation was removed.");
+                        } else {
+                          alert("Annotation was not removed.");
+                        }
+                      })
+                      .catch(error => {
+                        alert(JSON.stringify(error));
+                      });
                   }
                 });
               }}
@@ -743,23 +766,24 @@ class InstantExampleScreen extends Component<{}> {
   };
 
   onSuccess = qrData => {
-    const documentInfoUrl = qrData.data
+    const documentInfoUrl = qrData.data;
     fetch(documentInfoUrl)
       .then(response => {
         if (response.ok) {
           return response.json();
         }
-        throw new Error('Error message.');
-      }).then(data => {
-        this.props.navigation.popToTop();
-        PSPDFKit.presentInstant(data.serverUrl, data.jwt, {})
-      }).catch(e => {
-        console.log("failed to load ", url, e.message);
+        throw new Error("Error message.");
       })
-  }
+      .then(data => {
+        this.props.navigation.popToTop();
+        PSPDFKit.presentInstant(data.serverUrl, data.jwt, {});
+      })
+      .catch(e => {
+        console.log("failed to load ", url, e.message);
+      });
+  };
 
   render() {
-
     return (
       <View style={{ flex: 1 }}>
         {this.props.navigation.isFocused && (
@@ -767,8 +791,10 @@ class InstantExampleScreen extends Component<{}> {
             onRead={this.onSuccess}
             topContent={
               <Text style={styles.centerText}>
-                Scan the QR code from <Text style={styles.textBold}>pspdfkit.com/instant/demo</Text> to connect to the document shown in your browser.
-            </Text>
+                Scan the QR code from{" "}
+                <Text style={styles.textBold}>pspdfkit.com/instant/demo</Text>{" "}
+                to connect to the document shown in your browser.
+              </Text>
             }
           />
         )}
@@ -851,10 +877,10 @@ var styles = StyleSheet.create({
     flex: 1,
     fontSize: 18,
     padding: 32,
-    color: '#777',
+    color: "#777"
   },
   textBold: {
-    fontWeight: '500',
-    color: '#000',
+    fontWeight: "500",
+    color: "#000"
   }
 });

--- a/samples/Catalog/package.json
+++ b/samples/Catalog/package.json
@@ -1,6 +1,6 @@
 {
   "name": "Catalog",
-  "version": "1.24.8",
+  "version": "1.24.9",
   "private": true,
   "scripts": {
     "start": "node node_modules/react-native/local-cli/cli.js start"

